Added checks for missing sequence entries

Mailaender opened the pull request:
Closes #2283.

I experienced two crashes recently: 
- The sietch not having make animations while testing #4496.
- Deployed Tiberian Sun units not having death animations, a regression from https://github.com/OpenRA/OpenRA/pull/7687.

I extended the rule to also cover sequence definition on the actor from all traits (including overlays) and discovered some more inconsistencies in seldom used units. Also took some effort to clean things up.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Infantry/ScaredyCat.cs:
```diff
@@ -24,6 +24,8 @@ class ScaredyCatInfo : ITraitInfo
 		[Desc("Chance (out of 100) the unit has to enter panic mode when attacked.")]
 		public readonly int AttackPanicChance = 20;
 
+		[SequenceReference] public readonly string PanicSequencePrefix = "panic-";  // TODO: check the whole sequence with lint rules
```
Comment should go above the line.

Mailaender replied:
Updated.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Infantry/TakeCover.cs:
```diff
@@ -24,7 +24,7 @@ public class TakeCoverInfo : TurretedInfo
 		public readonly WVec ProneOffset = new WVec(85, 0, -171);
 
-		public readonly string ProneSequencePrefix = "prone-";
+		[SequenceReference] public readonly string ProneSequencePrefix = "prone-"; // TODO: check the whole sequence with lint rules
```
Same.

pchote commented on OpenRA.Mods.Common/Lint/CheckSequences.cs:
```diff
@@ -22,12 +22,36 @@ public void Run(Action<string> emitError, Action<string> emitWarning, Map map)
+							if (string.IsNullOrEmpty(sequence))
+								continue;
+
+							var renderInfo = actorInfo.Value.Traits.WithInterface<RenderSimpleInfo>().FirstOrDefault();
```
This needs to be using `RenderSpritesInfo.GetImage` if you want it to work in all cases.

Mailaender replied:
I don't have a valid race here though.

pchote replied:
You need to check every race, otherwise you may miss bogus sequences.

pchote commented on OpenRA.Mods.Common/Traits/Render/RenderSimple.cs:
```diff
@@ -41,11 +43,15 @@ public virtual int QuantizedBodyFacings(ActorInfo ai, SequenceProvider sequenceP
 
 	public class RenderSimple : RenderSprites, IAutoSelectionSize
 	{
+		RenderSimpleInfo info;
```
`readonly`.

pchote commented on OpenRA.Mods.Common/Traits/Render/RenderHarvester.cs:
```diff
@@ -16,7 +16,9 @@ namespace OpenRA.Mods.Common.Traits
 {
 	class RenderHarvesterInfo : RenderUnitInfo, Requires<HarvesterInfo>
 	{
+		/* [SequenceReference] TODO: lint rule does not support arrays */
```
We shouldn't be in a hurry to fix this, because it is a bit hacky.  We could perhaps do this with different body rendering sprites that are switched using upgrades.

Phrohdoh replied:
@pchote on the exact same TODO in another spot you said:
>You should fix this.

pchote replied:
Yup.  The comment here was first, and applied specifically to the context of the harvester sprites.  The other places are best represented with an array of images, so this is a problem for them.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithBuildingExplosion.cs:
```diff
@@ -19,6 +19,7 @@ namespace OpenRA.Mods.Common.Traits
 	class WithBuildingExplosionInfo : ITraitInfo, Requires<BuildingInfo>
 	{
 		[Desc("Explosion sequence name to use")]
+		/* [SequenceReference] TODO: not on the same actor */
```
You should extend the SequenceReferenceAttribute to accept an argument that overrides the sequence name (e.g. `[SequenceReference("explosion")]`)

pchote commented on OpenRA.Mods.Common/Traits/Render/WithInfantryBody.cs:
```diff
@@ -20,9 +20,15 @@ public class WithInfantryBodyInfo : ITraitInfo, IQuantizeBodyOrientationInfo, IR
+
+		[SequenceReference] public readonly string AttackSequence;
+
+		/* [SequenceReference] TODO: lint rule does not support arrays */
```
You should fix this.

reaperrr commented on OpenRA.Mods.Common/Traits/Render/WithInfantryBody.cs:
```diff
@@ -20,9 +20,15 @@ public class WithInfantryBodyInfo : ITraitInfo, IQuantizeBodyOrientationInfo, IR
+
+		[SequenceReference] public readonly string MoveSequence = "run";
+
+		[SequenceReference] public readonly string AttackSequence;
```
Why does this not define "shoot" as default?

Mailaender replied:
It will then complain that units like engineer don't define `shoot`.

Phrohdoh replied:
If that is the case, shouldn't we explicitly define `null`?

reaperrr replied:
I agree, that's what is usually done throughout the code.

penev92 replied:
Can we please refrain from actor- or mod-specific default values in the code?
That has caused enough messes so far.
It is much cleaner to have 10 actor types set their value, even if it is the same, and 2 that use the default empty value, then having to change or disable the default value that might end up being completely unrelated to your mod.
In general I think we should have as few "default" values hardcoded as possible.

Phrohdoh replied:
The engine default values are not going to go away.
Nothing presented here is actor or mod specific so I'm not sure what you're requesting.

pchote replied:
Defaulting this to null is fine, but it should be explicit for consistency.

penev92 replied:
@Mailaender's comment just above shows that in this particular case "shoot" is actor-specific.
Also you can refer to repairing/reloading to see "fix" being hardcoded as a default, which is pretty mod-specific.

Phrohdoh commented:
Needs a rebase.

Mailaender commented:
Rebased and extended this. Weapon sprites are still not covered so it won't detect https://github.com/OpenRA/OpenRA/issues/8020 yet.

abcdefg30 commented on OpenRA.Mods.Common/Lint/CheckSequences.cs:
```diff
@@ -19,15 +19,48 @@ class CheckSequences : ILintPass
+
+								var definitions = sequenceDefinitions.FirstOrDefault(n => n.Key == image.ToLowerInvariant());
+								if (definitions != null && !definitions.Value.Nodes.Any(n => n.Key.StartsWith(sequence))) // some sequences are suffixed
+								emitWarning("Sprite image {0} from actor {1} does not define sequence {2} from {3}".F(image, actorInfo.Value.Name, sequence, traitInfo));
```
Wrong indentation?

Mailaender replied:
Yep, thanks. Fixed.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Render/WithCrateBody.cs:
```diff
@@ -23,6 +23,10 @@ class WithCrateBodyInfo : ITraitInfo, Requires<RenderSpritesInfo>
 
+		[SequenceReference] public readonly string IdleSequence = "idle";
+		[SequenceReference] public readonly string WaterSequence;
+		[SequenceReference] public readonly string LandSequence;
```
Seems like you missed the defaults here?

penev92 replied:
Some times it's better not to set defaults in the code. That has caused *a few* messes.

pchote replied:
Convention is to explicitly specify null in these cases, even though they aren't needed.

Mailaender replied:
> Seems like you missed the defaults here?

It is redundant for crates that are not parachuted and don't need a landing sequence (for land and water) as they just appear and stay idle.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Render/WithDeathAnimation.cs:
```diff
@@ -17,15 +17,21 @@ namespace OpenRA.Mods.Common.Traits
+
 		[Desc("Sequence to play when this actor is crushed.")]
-		public readonly string CrushedSequence = "die-crushed";
+		[SequenceReference] public readonly string CrushedSequence;
```
Same here? (Or use `= null`)

Mailaender replied:
Some actors are uncrushable so this is optional. It will otherwise trigger a warning.

Mailaender commented:
Nullified everything and also found the next missing sequence crash regression. This time introduced by https://github.com/OpenRA/OpenRA/pull/8011. The thief is not supposed to take cover as he should run towards vehicles to capture them not duck and cover just to slow down and get killed. Therefore no prone sequences were defined so shooting THF will crash the game in bleed.

Mailaender commented:
I decided to split off the sequence error fixes from the refactorings to keep an overview.

abcdefg30 commented on mods/cnc/sequences/vehicles.yaml:
```diff
@@ -215,7 +215,7 @@ mlrs:
 		Start: 32
 		Facings: 32
-	turret-2:
+	aim:
```
This causes the `mlrs` to shoot even though it looks empty.

Mailaender replied:
I am a bit confused as msam has the same setup.

abcdefg30 replied:
The glitch is gone, but it looks like you don't use the empty sequence at all now.

Mailaender replied:
The problem is that the reload time is so short it actually does not reload but shoot shoot shoot shoot rockets until the aircraft falls to the ground.

abcdefg30 commented:
Otherwise :+1: / :white_check_mark:

Mailaender commented:
Updated the sequences a bit. The definitions seem to be completely bogus.

pchote commented:
The prospect of testing all these sequence changes (which nobody had noticed / files bugs about being bogus before) puts me off reviewing this.  Are they really necessary, or can they be deferred to a later PR?

Mailaender commented:
They will make the linter error and subsequently fail all builds so I had to include them. The make animation changes only refer to civilian buildings which have none, but the traits demand it. As they are not buildable, it never regressed to the players in the default mods.

penev92 commented:
The YAML part of the changes looks OK, but the is soooo much changed in the code :X

Mailaender commented:
It was just a grind task. Add `[SequenceReference]` before sprite sequence fields and extract them if they were hard-coded and then I had to fix the errors the linter spit out. The YAML part is the one that is hard to review because of the inheritance, not the source code.

obrakmann commented:
needs a rebase

Mailaender commented:
Will do so once https://github.com/OpenRA/OpenRA/pull/7982 has been merged to also cover that trait.

penev92 commented:
#7982 was merged.

pchote commented on OpenRA.Mods.Common/Lint/CheckSequences.cs:
```diff
@@ -19,15 +19,48 @@ class CheckSequences : ILintPass
+								if (renderInfo == null)
+									continue;
+
+								var image = renderInfo.GetImage(actorInfo.Value, map.SequenceProvider, null); // TODO: check all possible races
```
Forcing everything to use this image is a serious limitation of the feature - it means you need to break the configurability of traits like `WithSmoke`, and means that this cannot be used for other things that would benefit from it, like Projectiles.

Mailaender replied:
I now look up an image reference override so this won't need a hard-coded one anymore.

pchote replied:
There is no reason not to fix this now.  Query the races from the world actor, and then wrap this in a loop.

pchote commented on OpenRA.Mods.Common/Lint/CheckSequences.cs:
```diff
@@ -19,15 +19,48 @@ class CheckSequences : ILintPass
+
+								var sequenceReference = field.GetCustomAttributes<SequenceReferenceAttribute>(true).FirstOrDefault();
+								if (sequenceReference != null && !string.IsNullOrEmpty(sequenceReference.ImageOverride))
+									image = sequenceReference.ImageOverride;
```
You can change this `ImageOverride` to be the name of a field that contains the image for the class, and then fetch that value using reflection.

pchote commented:
There is another complication that this will need to handle: traits that can switch between different image definitions (e.g. `RenderSprite.RaceImages`, `RenderHarvester.ImagesByFullness`).  You will need to have a *Info interface that the lint pass can query to get an IEnumerable<string> of potential images, and then check the sequences for each.

Mailaender commented:
Filed https://github.com/OpenRA/OpenRA/pull/8223.

Mailaender committed:
check for missing sequence entries using lint

Mailaender commented:
Rebased.

pchote commented on OpenRA.Mods.Common/Lint/CheckSequences.cs:
```diff
@@ -19,15 +19,71 @@ class CheckSequences : ILintPass
 
 			foreach (var actorInfo in map.Rules.Actors)
+			{
 				foreach (var renderInfo in actorInfo.Value.Traits.WithInterface<RenderSimpleInfo>())
```
This needs to use `RenderSpritesInfo` and query the image name from `GetImage()`.  The render traits moved on from `renderInfo.Image ?? actorInfo.Value.Name` ages ago.

Mailaender replied:
This request is so complicated in itself and essentially unrelated that I moved it into pull request https://github.com/OpenRA/OpenRA/pull/8296.

pchote commented on OpenRA.Mods.Common/Lint/CheckSequences.cs:
```diff
@@ -19,15 +19,71 @@ class CheckSequences : ILintPass
+									var imageField = fields.FirstOrDefault(f => f.Name == sequenceReference.ImageReference);
+									if (imageField != null)
+									{
+										var imageOverride = LintExts.GetFieldValues(traitInfo, imageField, emitError).FirstOrDefault(); // TODO: check all
```
Likewise here.  Just loop over the following check for each result.

pchote commented on OpenRA.Mods.Common/Lint/CheckSequences.cs:
```diff
@@ -19,15 +19,71 @@ class CheckSequences : ILintPass
 						emitWarning("Sprite image {0} from actor {1} has no sequence definition.".F(image, actorInfo.Value.Name));
 				}
+
+				foreach (var traitInfo in actorInfo.Value.Traits.WithInterface<ITraitInfo>())
```
This has turned into a monster.  I suggest extracting some of the inner loops / blocks into their own methods.

Mailaender replied:
Done.

Mailaender committed:
document Country trait

Mailaender commented:
This now covers more cases (also race images and several images per actors) than I wanted to test in the first revision to be merged, but okay.

penev92 commented on OpenRA.Mods.Common/Traits/Parachutable.cs:
```diff
@@ -19,16 +19,20 @@ class ParachutableInfo : ITraitInfo
 		[Desc("If we land on invalid terrain for my actor type should we be killed?")]
 		public readonly bool KilledOnImpassableTerrain = true;
 
+		[Desc("Group where *CorposeSequence is looked up.")]
```
"corpose" -> "corpse"
And maybe `CorpseSequece`, instead of the * ?

penev92 commented on OpenRA.Mods.Common/Traits/Parachutable.cs:
```diff
@@ -19,16 +19,20 @@ class ParachutableInfo : ITraitInfo
 
 		public readonly string WaterImpactSound = null;
-		public readonly string WaterCorpseSequence = "small_splash";
+		[SequenceReference("CorpseSequenceCollection")] public readonly string WaterCorpseSequence = null;
```
Why are you removing the default value here?
While I don't like such magical default values, this will probably need an upgrade rule, and I really don't want this PR to get any bigger.

Mailaender replied:
If I leave the value which is ra mod exclusive, the linter will error for every other which does not have the sprite sequences.

penev92 commented:
This is now a monster PR :S
Please split some more stuff out into separate PRs to facilitate reviewing.
So far I've noticed the changes to `Parachutable`, however small (and only concerning the `Parachute` effect), need 2 upgrade rules, and include YAML changes in a lot of places. That would be a good PR on its own, I think.

Mailaender commented:
Depends on https://github.com/OpenRA/OpenRA/pull/8412 and https://github.com/OpenRA/OpenRA/issues/8296 now.

Mailaender commented:
Rebased.

This detected the following regressions from recent MiniYaml rule refactorings:
```
OpenRA.Utility(1,1): Warning: Sprite image doggie from actor doggie does not define sequence prefix prone- from field ProneSequencePrefix of OpenRA.Mods.Common.Traits.TakeCoverInfo
OpenRA.Utility(1,1): Warning: Sprite image cyborg from actor cyborg does not define sequence prefix prone- from field ProneSequencePrefix of OpenRA.Mods.Common.Traits.TakeCoverInfo
OpenRA.Utility(1,1): Warning: Sprite image cyc2 from actor cyc2 does not define sequence prefix prone- from field ProneSequencePrefix of OpenRA.Mods.Common.Traits.TakeCoverInfo

OpenRA.Utility(1,1): Warning: Sprite image dog from actor dog does not define sequence prefix prone- from field ProneSequencePrefix of OpenRA.Mods.Common.Traits.TakeCoverInfo
OpenRA.Utility(1,1): Warning: Sprite image zombie from actor zombie does not define sequence prefix prone- from field ProneSequencePrefix of OpenRA.Mods.Common.Traits.TakeCoverInfo
```

which proves my point: This is badly needed and I want to see it integrated and all rule simplifications rebased upon so future cleanups don't add new crashes.

obrakmann commented:
Got this when attacking with a zombie:

```
Exception of type `System.ArgumentNullException`: Argument cannot be null.
Parameter name: key
  at System.ThrowHelper.ThrowArgumentNullException (ExceptionArgument argument) [0x00000] in <filename unknown>:0 
  at System.Collections.Generic.Dictionary`2[System.String,OpenRA.Graphics.ISpriteSequence].FindEntry (System.String key) [0x00000] in <filename unknown>:0 
  at System.Collections.Generic.Dictionary`2[System.String,OpenRA.Graphics.ISpriteSequence].ContainsKey (System.String key) [0x00000] in <filename unknown>:0 
  at OpenRA.ReadOnlyDictionary`2[System.String,OpenRA.Graphics.ISpriteSequence].ContainsKey (System.String key) [0x00008] in /home/oliver/devel/openra/git/OpenRA.Game/Primitives/ReadOnlyDictionary.cs:67 
  at OpenRA.Graphics.SequenceProvider.HasSequence (System.String unitName, System.String sequenceName) [0x0003a] in /home/oliver/devel/openra/git/OpenRA.Game/Graphics/SequenceProvider.cs:81 
  at OpenRA.Graphics.Animation.HasSequence (System.String seq) [0x0000e] in /home/oliver/devel/openra/git/OpenRA.Game/Graphics/Animation.cs:169 
  at OpenRA.Mods.Common.Traits.WithInfantryBody.Attacking (OpenRA.Actor self, Target target) [0x00020] in /home/oliver/devel/openra/git/OpenRA.Mods.Common/Traits/Render/WithInfantryBody.cs:98 
  at OpenRA.Mods.Common.Traits.WithInfantryBody.Attacking (OpenRA.Actor self, Target target, OpenRA.Mods.Common.Traits.Armament a, OpenRA.Mods.Common.Traits.Barrel barrel) [0x00004] in /home/oliver/devel/openra/git/OpenRA.Mods.Common/Traits/Render/WithInfantryBody.cs:104 
```

Both ants and zombies are missing the proper `AttackSequence` in `WithInfantryBody`.

obrakmann commented:
Also, in TS every infantry attack I've tried causes the same crash as above. Isn't this lint check supposed to catch exactly this, or am I misunderstanding the purpose?

reaperrr commented:
What @obrakmann said, but :+1: once both issues have been adressed.

Mailaender commented:
> Isn't this lint check supposed to catch exactly this, or am I misunderstanding the purpose?

Probably a regression from https://github.com/OpenRA/OpenRA/pull/8171/ and other pull requests that were merged while this was ignored. I added more `[SequenceReference]` to bring those in line again.

If the default sequence is `null`, that is a valid rule. The game should either define a default or not crash in that case. I added a fix for the `ArgumentNullException`.

Mailaender commented:
Updated again. I should now also account for https://github.com/OpenRA/OpenRA/pull/7638.

obrakmann commented:
:+1:

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/ca1b0b41906f4b9bd1a0248a5813fd7ca77934b0%5E...ca1b0b41906f4b9bd1a0248a5813fd7ca77934b0)

penev92 commented:
Thanks! That was one hell of a cleanup :+1:

Mailaender commented:
Yep, I invested a lot of time, but I think it was worth it.

penev92 commented:
I just came across this [code](https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs#L48-L56).
With this PR there shouldn't be any more need for that exception silencing, right?
In fact, we can now afford to be less lenient about such issues and just crash.

pchote commented:
No.  We need that so that we can run the lint checks on travis, which doesn't have a copy of the assets.
